Optimize stream and add json output

AnderssonPeter opened the pull request:
Changed how streams are sent, before each logentry was sent in a separate stream, now they are grouped based on labels.
Also added support for sending json to loki, this allows more flexible log searching as you get fields for each log entry.

AnderssonPeter commented:
Hi i have 2 things I'm not 100% happy in this pull request.
1. The `forceLevelAsLabel` flag
2. How exceptions are encoded when sent over json

If you have any ideas how to handle those 2 better I'm happy to change/fix them.

mishamyte commented on Serilog.Sinks.Grafana.Loki.sln:
```diff
@@ -1,12 +1,22 @@
 ﻿
```
Could you, please, revert sln file?
I know that VS loves to regenerate sln file, but it's kinda annoying for me

AnderssonPeter replied:
I added some files to the solution file, i guess thats why you have changes in it, this way you can edit the readme.md file directly in Visual studio, but if you still want me to revert then sure i can do it.

mishamyte commented on src/Serilog.Sinks.Grafana.Loki/DictionaryComparer.cs:
```diff
@@ -0,0 +1,36 @@
+    {
+        public static DictionaryComparer<TKey, TValue> Instance { get; } = new DictionaryComparer<TKey, TValue>();
+
+        public bool Equals(IDictionary<TKey, TValue> x, IDictionary<TKey, TValue> y)
```
Should we implement here full Equality pattern?

Like

```csharp
            if (ReferenceEquals(x, y)) return true;
            if (ReferenceEquals(x, null)) return false;
            if (ReferenceEquals(y, null)) return false;
            if (x.GetType() != y.GetType()) return false;
            return x.Count == y.Count && !x.Except(y).Any();
```

AnderssonPeter replied:
Resolved with https://github.com/AnderssonPeter/serilog-sinks-grafana-loki/commit/b1e459e68ba34193282625d0603cf4b42f9b6549

mishamyte commented on src/Serilog.Sinks.Grafana.Loki/LokiJsonTextFormatter.cs:
```diff
@@ -0,0 +1,106 @@
+                .Where(pt => pt.Format != null);
+
+            // Better not to allocate an array in the 99.9% of cases where this is false
+            // ReSharper disable once PossibleMultipleEnumeration
```
Let's use `csharp[SuppressMessage("ReSharper", "PossibleMultipleEnumeration", Justification = "Reviewed")]` at class level

AnderssonPeter replied:
I don't have resharper so I'm not 100% sure i did it correctly but might be resolved with https://github.com/AnderssonPeter/serilog-sinks-grafana-loki/commit/549b20f70d85eba03bc92eacb9f6754620d72e48

mishamyte commented:
Hi, @AnderssonPeter!

**Thank you for your awesome contribution.**

1. Some time ago I took a decision to do a "level" label as always-present to use all Grafana level-highligting possibilities

In general case, it could parse it from log entry. But it depends a lot to log format. For example, Loki easily parses levels from ERR & DBG (which are output of template format), but can't deal with INF
![image](https://user-images.githubusercontent.com/5649656/106383251-97bb1d80-63cd-11eb-88fd-14755d77018c.png)

So, my point is to append this label (as fundamental all time). And it should be used for filtration (kinda native way, isn't it?)

But, in json format (with Loki's json parser) it causes to some duplication
![image](https://user-images.githubusercontent.com/5649656/106383354-11eba200-63ce-11eb-99aa-a6ba1e668477.png)

For this case I have a lil' bit more high level concept

2. As I see, you have defined some json format.
So it's kinda tricky point that format is predefined. I think, after your PR, we could go further and do something like a outputTemplate, but for json output. And in this case, we could give to user some tokens for building preferred format (for example, I won't add outputTemplate to json, but somebody could find this info useful) for json

3. Have done a small research and for now I don't see any way to do some beautiful encoding for exceptions

AnderssonPeter commented:
Thanks for your fast response!
1. This makes sense when encoding your logs as raw text (not json or logfmt)
2. If you are using the json format you more or less never want or need the level label as it will just cause the _extracted field, also allowing us to get rid of one extra label (If i understand loki fewer labels equals better performance).
The question is, could we somehow allow the `ITextFormatter` implementer decide if we should add a level label or not? That way we could get rid of the `forceLevelAsLabel` flag (its quite ugly)
3. Well i was pondering encoding them as json or a string but just didn't know what would be best json would be a bit powerful, but it could get nested fast if there where many inner exceptions, but sometimes the inner exception is the most interesting one and I'm not sure if it gets lost when calling `.ToString()`
But as a example we could have the following json
```json
{
  "exception": {
    "type": "System.Exception",
    "message": "....",
    "stackTrace": "....",
    "innerException": { ... }
  }
}
```
We could even handle the aggregate exception serialization.
4. I think it would be nice to have unit tests both for the serialization and also for the grouping of streams as this is a quite major change.

AnderssonPeter commented:
After looking at the output, maybe `messageTemplate` and `message` should be PascalCase instead? as most other fields seem to be PascalCase that .net core generates.

mishamyte commented:
@AnderssonPeter thank you for comments.

Sorry, have a work overload today and no time(((

Please, let me take a fresh look and I will give you the answers tommorow

mishamyte commented:
1. OK, it's clearly that LokiJsonTextFormatter doesn't need `level` label, cause it appends it by format. So, yes, it could be violation of single responsibility (partially). But could we exclude `level` while processing labels?
2. Nice idea, as for me it would be more useful. But I think, we should restrict exception nesting level while serializing. Or we could get a pretty bit matryoshka JSON of exception (I don't know, maybe some dudes write systems' error handling with a lot of nested exceptions)
3. What do u think about the format of output json? (yes, it is not the scope of this PR, but like a concept?)
4. Nice, will do it

mishamyte commented:
Also, nested exception could cause long labels name and it would be kinda strange.
Due to Loki spec of processing nested props.

`exception_innerException_innerException_message`

mishamyte commented:
For the PascalCase - agree. It's kinda native for Serilog and .NET in general

AnderssonPeter commented:
**Case**
Fixed PascalCase on all but level (i think loki requires us to have lower case here).

**Level label**
Added a property `ExcludeLevelLabel` on the `ILabelAwareTextFormatter` and if the formatter does not implement `ILabelAwareTextFormatter` then we always write the level label ensuring backwards compatibility.

**Exceptions**
I'm still a bit uncertain on what would be the best solution here
Pro:
Easy filtering based on type, message and so on without having to resort to regex.
Con:
Ugly field names `Exception_InnerException_InnerException_Message`.
Requires more storage to store.

**Json format**
I love that it allows us to easily filter based on a field without having to use regex, while still allowing us to define what field should be shown in the list (by clicking the eye)

**Solution**
I never got a response if you still want me to revert that one?

mishamyte commented:
Thank you, nicely done!

I consider that destructured exceptions are more useful here. For example, u could setup alerting more granular - for example, using concrete exception types for concrete service. So, as for me, it's a kinda good thing.

No, all right with the solution. Let's keep it.

AnderssonPeter commented:
Made a commit for destructured exceptions, but i don't have access to loki right now, could you verify that the output json is accepted by Loki?

AnderssonPeter commented:
Also it logs down to 3 destructured exceptions after that it just outputs strings.

mishamyte commented:
Yep, I will test it in the morning

AnderssonPeter commented:
I have tried it with one inner exception and it worked as intended!

mishamyte commented:
Have tested the last version.
As for me, all works as intended.

I'm ready to merge if there are no objections

AnderssonPeter commented:
Go for it :) and if possible could you create a beta nuget?

mishamyte commented:
Yes, I will do a beta soon. Also want to change versioning approach and go away from manual one

AnderssonPeter commented:
Any chance you could create a prerelease nuget package?

mishamyte commented:
Available in [v5.1.0-beta](https://github.com/mishamyte/serilog-sinks-grafana-loki/releases/tag/v5.1.0-beta)

Mario-Hofstaetter commented:
I have a stupid question @mishamyte : 
I am already using JSON logs with loki v2.1 (in our dev setup), using the [**RenderedCompactJsonFormatter**](https://github.com/serilog/serilog-formatting-compact) with success (see below), 

**what is changing with this PR?**
Is the default behaviour of the sink changing to match my current setup?

```csharp
.WriteTo.GrafanaLoki("http://localhost:3100",
        filtrationLabels: filtrationLabels, 
        filtrationMode:LokiLabelFiltrationMode.Include,
        textFormatter: new RenderedCompactJsonFormatter()
        )
    .CreateLogger();
```

`_m` contains the rendered log message for nice human readable view in Dashboards (as Loki / Grafana does not support something like this natively yet?)

```text
{Service="frontendservice"} | json | line_format "{{._m}}"
```

![grafana-explore-loki-parse-json](https://user-images.githubusercontent.com/33002073/107196124-8340db80-69f2-11eb-8552-c40dcf33db26.png)

![grafana-explore-loki-json-line-format](https://user-images.githubusercontent.com/33002073/107196112-7e7c2780-69f2-11eb-8689-84ea7b9c1cea.png)

mishamyte commented:
Hi, @Mario-Hofstaetter!

The default behavior won't change.

This PR implements some infrastructure optimization how data batch is built before sending to Loki.
And also it introduces out-of-box json formatter,  which can be used for serializing logs into JSON.

But I see that you have a success case of using another formatter. So I don't see any reason to change for you now

AnderssonPeter commented:
@Mario-Hofstaetter The reason for adding our own json formatter was mainly to avoid having the `level` as a label (fewer labels = better perfomance)
You can try it out with `new LokiJsonTextFormatter()`
